### PR TITLE
feat(protein-prospector): Add TMT converter for protein prospector

### PR DESCRIPTION
**Pull request by tonywu1999**
# Motivation and Context

In this [issue](https://github.com/Vitek-Lab/MSstatsPTM/issues/39), one of our users is requesting a new PTM+TMT converter for [Protein Prospector](https://prospector.ucsf.edu/prospector/mshome.htm).  In order to create this converter, we first need to create a TMT converter.

## Changes

- Add a new internal function `.cleanRawProteinProspector` to clean up the protein prospector output
- Create new `ProteinProspectortoMSstatsTMTFormat` function with basic functionality

## Testing

- Examples from .Rd files run successfully
- Locally tested on a 45MB dataset and it processed successfully.
- Added unit tests

## Checklist Before Requesting a Review
- [x] I have read the MSstats [contributing guidelines](https://github.com/Vitek-Lab/MSstatsConvert/blob/master/.github/CONTRIBUTING.md)
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: R/clean_ProteinProspector.R [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    )
+    
+    protein_prospector_input = melt(protein_prospector_input, 
+                                    measure.vars = channels, 
+                                    id.vars = setdiff(
+                                        colnames(protein_prospector_input), 
+                                        channels
+                                    ),
+                                    variable.name = "Channel", 
+                                    value.name = "Intensity"
```
Comment (mstaniak) [inline R/clean_ProteinProspector.R:30]: would it be useful to variable.factor = FALSE?

File: R/clean_ProteinProspector.R [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                                )
+    protein_prospector_input$Channel = .standardizeColnames(
+        protein_prospector_input$Channel
+    )
+    protein_prospector_input$Intensity = ifelse(
+        protein_prospector_input$Intensity == 0,
```
Comment (mstaniak) [inline R/clean_ProteinProspector.R]: Is it exactly 0 or should we use abs(Intensity) < 1e-8 for example?
Also the [, Intensity := ifelse(....)] would be better.
Moreover, there might be a fast ifelse in data.table

File: R/converters_ProteinProspectortoMSstatsTMTFormat.R [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        annotation, 
+        feature_columns,
+        remove_shared_peptides = useUniquePeptide,
+        remove_single_feature_proteins = removeProtein_with1Feature,
+        feature_cleaning = list(
+            remove_features_with_few_measurements = removeFewMeasurements,
+            summarize_multiple_psms = summaryforMultipleRows)
+        )
+    input = MSstatsConvert::MSstatsBalancedDesign(input, feature_columns,
+                                                  fix_missing = "zero_to_na")
```
Comment (mstaniak) [inline R/converters_ProteinProspectortoMSstatsTMTFormat.R:50]: Already happened in the cleaning function, can 0s appear again?
I'll check if this function uses my abs() < tolerance advice, btw

File: inst/tinytest/raw_data/ProteinProspector/Prospector_TotalTMT.txt
Comment (mstaniak) [inline inst/tinytest/raw_data/ProteinProspector/Prospector_TotalTMT.txt:1]: Other converters have unit tests, lets you this file for some simple test

**Review by mstaniak — DISMISSED** [review]

File: R/clean_ProteinProspector.R (hunk 2 above)
Comment (mstaniak) [reply R/clean_ProteinProspector.R]: Update: MSstatsBalancedDesign does the same. It seems that input files should have exact 0s as missing values. I wonder if there's any case when this fails. Probably not

File: R/clean_ProteinProspector.R (hunk 1 above)
Comment (tonywu1999) [reply R/clean_ProteinProspector.R:30]: I'm not sure if there'll be any added benefit.  

I also tested on the dataset provided to see if there's any difference in space usage, and using factors reduced the memory usage of the channels column by ~50% (8112 bytes -> 4680 bytes of space stored for the Channel column).  This seems to be supported by this [post](https://stackoverflow.com/questions/74608288/in-r-do-factors-somehow-save-space) since there's many duplicates.

File: R/clean_ProteinProspector.R (hunk 2 above)
Comment (tonywu1999) [reply R/clean_ProteinProspector.R]: Oh yeah you're right, balanced design does handle this case.  I'll remove this code here.

Comment (tonywu1999) [reply R/clean_ProteinProspector.R]: I'll need to double check with the end user on what criteria warrants a missing value. I also looked through the larger input dataset provided and saw some negative numbers too, which is interesting.

File: inst/tinytest/raw_data/ProteinProspector/Prospector_TotalTMT.txt
Comment (tonywu1999) [reply inst/tinytest/raw_data/ProteinProspector/Prospector_TotalTMT.txt:1]: Yup, I'll add a unit test

File: R/clean_ProteinProspector.R (hunk 1 above)
Comment (mstaniak) [reply R/clean_ProteinProspector.R:30]: right, the integer representation helps, let's leave it as it is

File: R/converters_ProteinProspectortoMSstatsTMTFormat.R (hunk 3 above)
Comment (tonywu1999) [reply R/converters_ProteinProspectortoMSstatsTMTFormat.R:50]: I removed the `ifelse` intensity check from the clean function since this seems to do the same thing too.  But to answer your question, yes, it looks like Preprocess causes 0 to appear again too. 

https://github.com/Vitek-Lab/MSstatsConvert/blob/devel/R/utils_clean_features.R#L293-L305

And then those get converted to NA in MSstatsBalancedDesign.

**Comment by tonywu1999** [comment]
TODO:

- ~~Determine how to handle 0s, negative values, and values between 0 and 1~~
- ~~Add unit tests on 0-1, 0s, & negative value handling~~

File: R/converters_ProteinProspectortoMSstatsTMTFormat.R (hunk 3 above)
Comment (tonywu1999) [reply R/converters_ProteinProspectortoMSstatsTMTFormat.R:50]: ^additional observation: It seems that negative values slip through the cracks here, but they later turn into NAs when applying log transformation.

File: R/clean_ProteinProspector.R (hunk 2 above)
Comment (tonywu1999) [reply R/clean_ProteinProspector.R]: Ok I confirmed 0s and negative values are missing values

**Comment by mstaniak** [comment]
Hi @tonywu1999 , adding 1:n() to PSM is meant to help retain the identity of PSMs when melting, right?
The single underscore will/might be a problem for dataProcess(), "_[PSM ID]" should be removed after aggregating PSMs.
Also the := will work a bit better, data.table might have a faster ifelse, too

**Comment by tonywu1999** [comment]
@mstaniak 

> adding 1:n() to PSM is meant to help retain the identity of PSMs when melting, right? 

Yes, and I believe retaining the identity of PSMs is needed when aggregating them [here](https://github.com/Vitek-Lab/MSstatsConvert/blob/devel/R/utils_clean_features.R#L157-L158) within the same run/sequence/charge combination.  

> The single underscore will/might be a problem for dataProcess(), "_[PSM ID]" should be removed after aggregating PSMs.

It looks like the single underscore from adding `1:n` is removed [here](https://github.com/Vitek-Lab/MSstatsConvert/blob/devel/R/utils_clean_features.R#L192-L193).  

> Also the := will work a bit better, data.table might have a faster ifelse, too

I ended up updating the code to use := when adding the PSM column
